src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+            }
+            m_nextActivation += (1.0f / m_effectParams->rate);
+        }
+
+    }
+}
+
+void ParticleEffect::updatePosition(Particle *p, float dt, float attenuationFactor) {
+    p->velocity *= attenuationFactor;
+    for (int i = 0; i < (int)m_effectParams->accelerators->size(); i++) {
+        p->velocity += m_effectParams->accelerators->at(i) * dt;
+    }
+    p->position += p->velocity * dt; 
+}
+
+int ParticleEffect::getAttenuations(float dt) {
+    if (m_effectParams->attenuation == 1.0f) {
+        return 0;
+    }
+    else {
+        int attenuations = 0;
+        while (m_life > m_nextAttenuation) {
+            attenuations += 1;
+            m_nextAttenuation += m_attenuationRate;
+        }
+        return attenuations;
+    }
+}
+
+ //Remove from specific location
+void ParticleEffect::removeActiveParticle(int i) {
+    m_activeMap.erase(m_activeMap.begin() + i);
+    m_activeParticlePositions->erase(m_activeParticlePositions->begin() + i);
+    m_activeParticleOrientationIDs->erase(m_activeParticleOrientationIDs->begin() + i);
+}
+
+//Will always add to end of vector
+void ParticleEffect::addActiveParticle(int i) {
+    m_activeMap.push_back(i);
+    m_activeParticlePositions->push_back(m_particles[i]->position);
+    m_activeParticleOrientationIDs->push_back(m_particles[i]->orientationID);
+}
+
+void ParticleEffect::sphereMotion(Particle* p) {
+    if (!m_effectParams->randomDistribution) {
+        float phi = glm::golden_ratio<float>();
+        float z = 1 - (2 * float(p->i) / float(m_effectParams->n - 1));
+        float radius = sqrt(1 - z * z);
+        float theta = 2 * glm::pi<float>() * (2 - phi) * float(p->i);
+        p->velocity = glm::vec3(radius * cos(theta), radius * sin(theta), z);
+    }
+    else {
+        p->velocity = normalize(glm::ballRand(1.0f));
+    }
+}
+
+void ParticleEffect::diskMotion(Particle* p) {
+    if (!m_effectParams->randomDistribution) {
+        float theta = m_effectParams->angle * (p->i / (float)m_effectParams->n);
+        p->velocity = glm::vec3(cos(theta), 0.0f, sin(theta));
+    }
+    else {
+        float theta = getRandom(-1 * m_effectParams->angle / 2.0f, m_effectParams->angle / 2);
+        p->velocity = glm::vec3(cos(theta), 0.0f, sin(theta));
+    }
+}
+
+void ParticleEffect::coneMotion(Particle* p) {
+    
+    glm::vec3 normalizedDirection = normalize(m_direction);
+    glm::vec3 u = getU(normalizedDirection);
+    glm::vec3 v = glm::cross(u, normalizedDirection);
+    float angle = m_effectParams->angle;
+    float phi = getRandom(-1 * glm::pi<float>(), glm::pi<float>());
+    float z = getRandom(cos(angle), 1);
+    float theta = acosf(z);
+    p->velocity = sin(theta) * (cos(phi) * u + sin(phi) * v) + cos(theta) * normalizedDirection;
+}
+
+glm::vec3 ParticleEffect::getU(glm::vec3 directionNorm) {
+    glm::vec3 randVec = glm::ballRand(1.0f);
+    glm::vec3 u = normalize(cross(directionNorm, randVec));
+    while (u == glm::vec3(0.0f)) {
```

Comment:
Why is this a while loop?

Reply:
Just in case u is ever glm:vec3(0,0,0). It's unlikely, but you never know.